index.bs

```diff
@@ -787,14 +718,595 @@ the following list:
+ 1. Let |abortAlgorithm| be an action that [=aborts=] |stream| with |reason|, given |reason|.
+ 1. Let |writable| be a [=new=] {{WritableStream}}.
+ 1. [=WritableStream/Set up=] |writable| with |writeAlgorithm|, |closeAlgorithm|, and
+    |abortAlgorithm|.
+ 1. Set |stream|'s [=WebSocketStream/readable stream=] to |readable|.
+ 1. Set |stream|'s [=WebSocketStream/writable stream=] to |writable|.
+ 1. [=Resolve=] |stream|'s [=WebSocketStream/opened promise=] with {{WebSocketOpenInfo}} «[
+    "{{WebSocketOpenInfo/extensions}}" → |extensions|, "{{WebSocketOpenInfo/protocol}}" →
+    |protocol|, "{{WebSocketOpenInfo/readable}}" → |readable|, "{{WebSocketOpenInfo/writable}}" →
+    |writable| ]».
+
+</div>
+
+<hr>
+
+<div algorithm="handle a WebSocketStream message">
+
+When [=a WebSocket message has been received=] for a {{WebSocketStream}} |stream| with type |type|
+and data |data|, the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to follow these steps: [[!WSP]]
+
+ 1. If [=WebSocket/ready state=] is not {{WebSocket/OPEN}} (1), then return.
+ 1. Let |chunk| be determined by switching on |type|:
+
+  <dl class="switch">
+   : |type| indicates that the data is Text
+   :: a new {{DOMString}} containing |data|
+
+   : |type| indicates that the data is Binary
+   :: a new {{Uint8Array}} object, created in the [=relevant Realm=] of the {{WebSocketStream}}
+      object, whose contents are |data|
+  </dl>
+
+ 1. [=ReadableStream/Enqueue=] |chunk| into |stream|'s [=WebSocketStream/readable stream=].
+ 1. Apply backpressure to the WebSocket.
+
+ <p class=note>Applying backpressure <span class=allow-2119>should</span> result in no new WebSocket
+ messages being handled until the backpressure is released. After an implementation-defined amount
+ of data is buffered, the client will stop reading from the underlying network connection resulting
+ in network-layer backpressure being applied.
+</div>
+
+<hr>
+
+When [=the WebSocket closing handshake is started=] for a {{WebSocketStream}}
+|stream|, the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to change the |stream|'s [=WebSocketStream/ready state=] to
+{{WebSocket/CLOSING}} if it is not already. [[!WSP]]
+
+<hr>
+
+<div algorithm="handle WebSocketStream connection close">
+
+When [=the WebSocket connection is closed=] for a {{WebSocketStream}} |stream|, possibly
+[=cleanly=], the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to run the following substeps:
+
+ 1. Change the [=WebSocketStream/ready state=] to {{WebSocket/CLOSED}} (3).
+ 1. If |stream|'s [=WebSocketStream/handshake aborted=] is true, then return.
+ 1. If |stream|'s [=WebSocketStream/was ever connected=] is false, then [=reject=] |stream|'s
+    [=opened promise=] with a [=new=] WebSocketError.
```

Comment:
The closeCode and reason are not stated for this WebSocketError, and no defaults are given in the class definition. Probably you want to add defaults in the class definition?